retrofit-mock/src/test/java/retrofit/MockRetrofitTest.java

```diff
@@ -0,0 +1,327 @@
+    mockRetrofit.setDelay(100, MILLISECONDS);
+    mockRetrofit.setVariancePercentage(0);
+    mockRetrofit.setErrorPercentage(0);
+
+    IOException failure = new FileNotFoundException("Oh noes");
+    Call<Object> call = mockRetrofit.newFailureCall(failure);
+
+    long startNanos = System.nanoTime();
+    try {
+      call.execute();
+      fail();
+    } catch (IOException e) {
+      long tookMs = TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos);
+      assertThat(tookMs).isGreaterThanOrEqualTo(100);
+      assertThat(e).isSameAs(failure);
+    }
+  }
+
+  @Test public void asyncFailureCalledAfterDelay() throws InterruptedException {
+    mockRetrofit.setDelay(100, MILLISECONDS);
+    mockRetrofit.setVariancePercentage(0);
+    mockRetrofit.setErrorPercentage(0);
+
+    IOException failure = new FileNotFoundException("Oh noes");
+    Call<Object> call = mockRetrofit.newFailureCall(failure);
+
+    final AtomicLong tookMs = new AtomicLong();
+    final AtomicReference<Throwable> failureRef = new AtomicReference<>();
+    final CountDownLatch latch = new CountDownLatch(1);
+    final long startNanos = System.nanoTime();
+    call.enqueue(new Callback<Object>() {
+      @Override public void success(Response<Object> response) {
+        throw new AssertionError();
+      }
+
+      @Override public void failure(Throwable t) {
+        tookMs.set(TimeUnit.NANOSECONDS.toMillis(System.nanoTime() - startNanos));
+        failureRef.set(t);
+        latch.countDown();
+      }
+    });
+    assertTrue(latch.await(1, SECONDS));
+
+    assertThat(tookMs.get()).isGreaterThanOrEqualTo(100);
+    assertThat(failureRef.get()).isSameAs(failure);
+  }
+
+  @Test public void syncCanBeCanceled() throws IOException {
+    mockRetrofit.setDelay(10, SECONDS);
+    mockRetrofit.setVariancePercentage(0);
+    mockRetrofit.setErrorPercentage(0);
+
+    final Call<String> call = mockRetrofit.newSuccessCall("Hi");
+
+    new Thread(new Runnable() {
+      @Override public void run() {
+        try {
+          Thread.sleep(100);
+          call.cancel();
+        } catch (InterruptedException ignored) {
+        }
+      }
+    }).start();
+
+    try {
+      call.execute();
+      fail();
+    } catch (InterruptedIOException e) {
+      assertThat(e).hasMessage("canceled");
+    }
+  }
+
+  @Test public void asyncCanBeCanceled() throws InterruptedException {
+    mockRetrofit.setDelay(100, MILLISECONDS);
+    mockRetrofit.setVariancePercentage(0);
+    mockRetrofit.setErrorPercentage(0);
+
+    final Call<String> call = mockRetrofit.newSuccessCall("Hi");
+
+    final CountDownLatch latch = new CountDownLatch(1);
+    call.enqueue(new Callback<String>() {
+      @Override public void success(Response<String> response) {
+        latch.countDown();
+      }
+
+      @Override public void failure(Throwable t) {
+        latch.countDown();
+      }
+    });
+
+    call.cancel();
+
+    assertFalse(latch.await(1, SECONDS));
```

Comment:
In OkHttp cancel triggers the failure callback. I think we might want retrofit to do likewise.